### ISSUE TITLE
Spatial filtering: Working copy checkout is slow to skip missing features

### ISSUE THREAD

**Issue by olsen232** [koordinates/kart#561]
This could depend on the size and complexity of features, but testing locally, Kart is currently 70% slower at skipping features that are missing+promised compared to finding and writing a feature that is present locally. This means that creating a spatially filtered clone may or may not be faster than a regular clone, depending on if the time saved by sending less features over the network is counteracted by more time spent trying to read those missing features.

The slowness could be due to any or all of the following:
- raising and catching more exceptions
- git ODB looking in more places for the missing features before finally accepting they are missing
- something else?

**Comment by olsen232** [comment]
Leaning towards the second possibility, which unfortunately means there is no obvious fix - maybe we can delve deep enough into libgit2 to optimise it somehow.
I've been running the following git command to merely print the OIDs of all the objects and verify whether or not they are missing - I don't think it actually loads the data from each object. The pv bit at the end tells me how many lines git is printing per second.
(I'm using a git command directly since it avoids any inefficiencies of going between C and Python repeatedly, including inefficiencies I have added myself. No guarantee that git is doing it as efficiently as possible but it is the best indication we have).

`time git rev-list HEAD --objects --missing=allow-any | pv --line-mode --rate > /dev/null`

- If I run this on a Kart repo where all of the objects were fetched during clone, it runs at 200k to 310k objects per second.
- If I run this on a Kart repo where no objects were fetched during clone - all objects are missing+promised - it runs at 300 to 400 trees per second. Since each tree contains 50 - 60 missing objects in this repository, this is equivalent to running at around 20k objects per second - much slower than the first example.

This may not be a completely fair comparison since from within Kart, it's not enough know all the OIDs and whether they are missing or not - we need to know all the OIDs (without which we can't look up the contents), whether they are missing, *and* the actual contents of any non-missing ones. Looking up the contents would obviously slow down the first example but not the second, since the second has no blob content. However, a full working copy checkout in the first repo completes quicker than this command runs in the second anyway - about 30k features per second. 

In conclusion it seems difficult for us to make working copy checkout in the second repo go as fast as the first repo (30k FPS) when we can't even ask git to print all the non-missing feature OIDs for that fast in the second repo (only 20k FPS), and that knowing the OIDs and if they are missing is a subset of the problem of checking out the working copy.

The good news is that switching between commits often doesn't involve a full working copy checkout, and so will generally be much faster (relatively few features have changed, and relatively few features outside the spatial filter have changed).

**Comment by olsen232** [comment]
Another potential fix would be not to fetch trees if they contain no features - this involves working on the spatial filter git extension, so that we prune the tree as we send it, and work on Kart so that we don't assume we have every single tree and a pointer to every single feature (even if that feature is missing). A simple example, the true `feature-count` of a dataset would no longer be known locally. Or more complex, the call to merge_trees would no longer work if there were missing trees that had both changed that needed merging - the trees would have to be fetched first. (Right now we only need to fetch the features that turn out to be conflicting after we perform the entire merge).

 But, it would speed up working copy checkout as entire trees could be detected as missing and skipped in a single step, instead of skipping each feature in them one by one.
